File: Week_03/java/src/combinations/_216_CombinationSumIII.java
package combinations;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**
 * 第一遍：2020/09/10周三 ✅
 * 第二遍：2020/09/10周三
 * 第三遍：2020/06/14周日
 * 第四遍：2020/06/28周日
 * 佩服佩服，多多练习。
 * 这两种方法都挺妙的，从不同的角度去考虑问题。
 * 纵：考虑每一个元素选还是不选
 * 横：一次性考虑多个元素
 */
public class _216_CombinationSumIII {
    // 时间O(2^10) == O(1) 空间 O(1)
    static class 纵 {
        List<List<Integer>> ret = new LinkedList<>();
        LinkedList<Integer> path = new LinkedList<>();

        public List<List<Integer>> combinationSum3(int k, int n) {
            dfs(1, n, k);
            return ret;
        }

        void dfs(int i, int n, int k) {
            if (n == 0 && k == 0) {
                ret.add(new ArrayList<>(path));
                return;
            }
            if (k <= 0 || i > n || i == 10) return;
            path.offerLast(i);
            dfs(i + 1, n - i, k - 1);
            path.pollLast();
            dfs(i + 1, n, k);
        }
    }

    // 时间O(10^10) == O(1) 空间 O(1)
    static class 横 {
        List<List<Integer>> ret = new LinkedList<>();
        LinkedList<Integer> path = new LinkedList<>();

        public List<List<Integer>> combinationSum3(int k, int n) {
            dfs(1, n, k);
            return ret;
        }

        void dfs(int start, int n, int k) {
            if (n == 0 && k == 0) {
                ret.add(new ArrayList<>(path));
                return;
            }
            if (k <= 0) return;
            for (int i = start; i <= 9; i++) {
                if (i > n) return;
                path.offerLast(i);
                dfs(i + 1, n - i, k - 1);
                path.pollLast();
            }
        }
    }
}

File: Week_03/cpp/src/combinations/_216_CombinationSumIII.cpp
#include <vector>
#include <iostream>

using namespace std;

class 横
{
public:
    vector<vector<int>> ret;
    vector<int> path;
    // 时间O(1) 空间O(1) 通过在候选数组中，用一个数丢弃一个数的方式
    // 来保证组合中不存在重复的数字同时也保证了不存在重复组合,因为候选数组中不存在重复数字。
    vector<vector<int>> combinationSum3(int k, int n)
    {
        dfs(1, n, k);
        return ret;
    }
    void dfs(int start, int n, int k)
    {
        if (k == 0 && n == 0)
        {
            ret.push_back(path);
            return;
        }
        k--;
        for (int i = start; i <= 9; i++)
        {
            if (i > n)
                return;
            path.push_back(i);
            dfs(i + 1, n - i, k);
            path.pop_back();
        }
    }
};
class 纵
{
public:
    vector<vector<int>> ret;
    vector<int> path;
    vector<vector<int>> combinationSum3(int k, int n)
    {
        dfs(1, n, k);
        return ret;
    }
    void dfs(int i, int n, int k)
    {
        if (k == 0 && n == 0)
        {
            ret.push_back(path);
            return;
        }
        if (k == 0 || i > 9 || i > n)
            return;
        path.push_back(i);
        dfs(i + 1, n - i, k - 1);
        path.pop_back();
        dfs(i + 1, n, k);
    }
};